app/main/authorisation.py

```diff
@@ -1,57 +1,90 @@
-from flask import abort, current_app, g
+from abc import ABC, abstractmethod
 
 
-def check_authorised() -> tuple[tuple[str], dict[str]]:
-    """Checks that the user is authorized to submit.
+class AuthBase(ABC):
+    """Auth class ABC. Classes that inherit must implement a constructor, organisations and auth_dict methods."""
 
-    Returns any LAs, places, and fund types that the user is authorized to submit for, otherwise aborts and redirects
-    to 401 (unauthorised) page.
+    @abstractmethod
+    def __init__(self, *args):
+        pass
 
-    :return: the LAs as a tuple, and a dictionary with both the place_names and fund_types
-    """
-    local_authorities, place_names, fund_types = get_local_authority_and_place_names_and_fund_types(g.user.email)
-    if local_authorities is None or place_names is None or fund_types is None:
-        current_app.logger.error(
-            f"User: {g.user.email} has not been assigned any local authorities and/or places and/or fund types"
-        )
-        abort(401)  # unauthorized
-    current_app.logger.info(
-        f"User: {g.user.email} from {', '.join(local_authorities)} is authorised for places: {', '.join(place_names)}"
-        f"and fund types: {', '.join(fund_types)}"
-    )
-    return local_authorities, {"Place Names": place_names, "Fund Types": fund_types}
-
-
-def get_local_authority_and_place_names_and_fund_types(
-    user_email: str,
-) -> tuple[tuple[str] | None, tuple[str] | None, tuple[str] | None]:
+    @abstractmethod
+    def get_organisations(self) -> tuple[str, ...]:
+        """Return organisations associated with this level of authorisation."""
+        pass
+
+    @abstractmethod
+    def get_auth_dict(self) -> dict:
+        """Return other details associated with this authorisation."""
+        pass
+
+
+class TFAuth(AuthBase):
+    """A Towns Fund Auth Class"""
+
+    local_authorities: tuple[str, ...]
+    place_names: tuple[str, ...]
+    fund_types: tuple[str, ...]
+
+    def __init__(self, local_authorities: tuple[str, ...], place_names: tuple[str, ...], fund_types: tuple[str, ...]):
+        self.local_authorities = local_authorities
+        self.place_names = place_names
+        self.fund_types = fund_types
+
+    def get_organisations(self) -> tuple[str, ...]:
+        return self.local_authorities
+
+    def get_auth_dict(self) -> dict:
+        return {"Place Names": self.place_names, "Fund Types": self.fund_types}
+
+
+class AuthMapping:
+    """Encapsulates an email mapping dictionary. Allows lookup of an email address."""
+
+    _auth_class: type[AuthBase]
+    _mapping: dict[str, AuthBase]
+
+    def __init__(self, auth_class: type[AuthBase], mapping: dict[str, tuple[tuple[str, ...], ...]]):
+        self._auth_class = auth_class
+        self._mapping = {email: auth_class(*auth_details) for email, auth_details in mapping.items()}
```

Comment:
It's not super clear what ```_auth_class``` is glancing at this, and line 49 is not especially intuitive either. Maybe some comments to explain this class a bit more, or at least something to indicate what's going on in the instantiation of the mapping?

Reply:
good shout - I've added docstrings to the constructor now